### PR TITLE
Feat/drop: add dropout methods

### PR DESCRIPTION
**Pull request by Nhogg**
# Image augmentation primitives in the Grain pipeline

## Summary

Adds three per-sample image augmentations to the Grain data pipeline, each gated by an independent probability exposed on the tyro Config:

* **`patch_occlude`** - zero out a random rectangular patch in each image view
* **`image_view_drop`** - zero out entire views per `(sample, view)` and flip the corresponding `pad_mask_dict` entry to `False`
* **`image_key_shuffle`** - permute which view occupies which image-key slot, per sample, carrying `pad_mask_dict` along with the view

All three live in `crossformer/data/grain/transforms.py`, are wired into `GrainDataFactory` in `crossformer/data/grain/loader.py`, and surfaced via `Config` in `scripts/train/xflow.py`.

## Pipeline placement

The transforms sit in the main-process CPU stage, between `ThreadPrefetchIterDataset` and `np2jax` (see `loader.py:395-430`). This is intentional:

* **Before `np2jax`**: Tensors are still numpy, so we can mutate nested dicts cheaply and avoid round-tripping through JAX.
* **After `ThreadPrefetchIterDataset`**: The worker pool has already delivered batched samples, so the per-sample randomness runs once per batch rather than per trajectory.
* **Cost**: Each op is gated with `if train and self.<prob> > 0` so eval and default configs are no-ops (zero runtime cost).

Eval uses the same factory but with `prob=0` unless the user explicitly enables; the pass-through in the eval factory call mirrors train so that debug sweeps with nonzero prob also exercise the eval path if desired.

## Data-layout notes

At this stage the observation dict is still nested:

* `step["observation"]["image"][view]` -  shape `(B, T, H, W, C)` or `(B, H, W, C)` depending on upstream chunking.
* `step["observation"]["pad_mask_dict"]["image"][view]` - bool mask with matching leading dims.

The later `compatibility` transform flattens nested image keys to `image_<view>` post-GPU; we operate before that flattening so `images[view]` and `pmd[view]` are both directly indexable.

## `patch_occlude`

Per `(sample, timestep, view)` with probability `prob`, zero out a rectangular region with area fraction in `[min_frac, max_frac]` of `H*W` and aspect ratio in `[0.5, 2.0]`.

**Shape handling**: The array may arrive as 4D `(B, H, W, C)` or 5D `(B, T, H, W, C)`. `patch_one` inserts a time axis if missing, runs the masked write, and squeezes back. This is why the code looks like `squeezed = a.ndim == 4; if squeezed: a = a[:, None]; ...; return out[:, 0] if squeezed else out` -  defensive in case upstream chunking moves.

**Mask construction**: `y_in & x_in` are broadcast together to form a `(B, T, H, W)` boolean patch mask, which is then broadcast against the channel axis via `patch[..., None]`. All-numpy, no Python loops over samples.

## `image_view_drop`

Per `(sample, view)` with probability `prob`, zero the view across all timesteps and set `pad_mask_dict["image"][view]` to `False` for those samples.

**Views drop independently**: There is no "at least one survives" rescue. With K views at prob p, the probability that a sample loses all views is `p**K`, which is negligible for typical (p=0.1–0.3, K=3). If this ever matters, rescue logic belongs in this function, not downstream.

**Mask semantics**: The existing `pmd_img[v]` may already be `False` (upstream padding); the new mask is `existing & ~drop`, never flipping a previously-False entry to True.

## `image_key_shuffle`

Per sample with probability `prob`, permute which view occupies which image-key slot. Pad masks travel with the view.

**Why it's vectorized the way it is**:
* `rng.random(b) < prob` - per-sample Bernoulli hit.
* `np.argsort(rng.random((b, K)), axis=1)` - vectorized permutation trick. Sorting uniform noise gives a uniformly random permutation per row, with no loop over samples.
* `np.where(shuffled[:, None], random_perm, np.arange(K))` - unhit samples use the identity `[0..K-1]` permutation, so the gather below is a no-op for them.
* `imgs = np.stack([images[v] for v in views], axis=1)` → `(B, K, T, H, W, C)`. Stacking on axis 1 means the K axis is adjacent to B, which is what the advanced-indexing gather expects.
* **The gather**: `imgs[bi, perm]` where `bi = np.arange(b)[:, None]` is `(B, 1)` and `perm` is `(B, K)`. Numpy broadcasts `bi` across K so the paired index at `(row, col)` is `(row, perm[row, col])` - i.e. for each sample, pull view `perm[b, k]` into slot `k`. Without the `[:, None]` on `bi`, numpy would pair `bi[i]` with `perm[i, :]` along the same axis and the gather would not express per-sample permutations.

**Dict rebuild**: New dicts are keyed by the original `views` list in original order - the permutation changes what data lives at each key, not the key names themselves. This keeps downstream tokenizer wiring oblivious to the shuffle.

## `proprio_sample_drop`
Per sample with probability `prob`, zero **all** proprio subkeys and flip **all** `pmd` entries to False. One Bernoulli hit per sample, applied uniformly across every subkey.

**Semantics - total sensor loss**: this models the whole proprio sensor stack going away for a training sample. Per-subkey independence (e.g. joint present but pose missing) is not handled here, that is the job of an in-model `proprio_token_drop` operating at token granularity. Keeping coarse sample-level drop in Grain and the fine per-key drop in model lets each augmentation run at the stage that fits it best: Grain is CPU-worker overlapped and well suited to whole-tensor ops; token drop needs the JAX rng stream and happens post-tokenization anyway.

**Data layout note**: at this pipeline stage, `proprio` is a nested dict keyed by body-part (`joints`, `pose`, `gripper`, ...), with each value shaped `(B, T, D)` for flat subkeys or `(B, T, ...)` for structured ones (e.g. `pose` as `(B, T, 2, 3)`). The `(x.ndim - 1)` reshape in `zero()` and `new_mask()` handles both uniformly.

## Why not on GPU

`image_key_shuffle` could in principle run post-`np2jax` as a `jnp.take`/gather under `jit`. We keep it in Grain because:

1. Grain runs in the CPU worker pool, overlapping with GPU step time. For a GPU-bound training loop the shuffle cost is hidden behind compute.
2. Running on GPU would require threading another PRNG stream through the model; keeping image-level ops in Grain leaves model code with only the one rng stream needed for in-model token-level drops.
3. The work is bandwidth-dominated, not compute-dominated, so the "GPU is faster" intuition doesn't really apply to a gather over image tensors.

## Config surface

Added to `scripts/train/xflow.py`:

```python
patch_prob: float = 0.0
patch_min_frac: float = 0.05
patch_max_frac: float = 0.5
view_drop_prob: float = 0.0
image_shuffle_prob: float = 0.0
```
Each component flows through the following pipeline:
`Config` → `GrainDataFactory(...)` kwarg → `self.<field>` → gated `random_map` in `.make()`. 

> [!IMPORTANT]
> Defaults of **0.0** make every gate inert, ensuring this PR is a no-op behavior change for users who do not pass new flags.

---

## Test Plan

- [x] **Baseline Parity**: `uv run scripts/train/xflow.py` with all probs at 0.
- [ ] **Patch Occlusion**: `--patch-prob 0.5` - verify patches appear in a sample batch.
- [x] **View Dropping**: `--view-drop-prob 0.5` - verify zeroed views and corresponding `pad_mask_dict` entries flipped to `False`.
- [x] **Key Shuffling**: `--image-shuffle-prob 1.0` - verify per-sample multiset preserved, masks travel with views, and ~5/6 of samples end up non-identity (with $K=3$).
- [x] **Integration**: All three enabled together - verify no shape errors through tokenizer/model forward pass.

### CODE REVIEW

**Comment by Nhogg** [comment]
# Method Verification
Ran each primitive in isolation at `prob=1.0` on a real batch (`B=256`, 4 camera views) to verify correctness. All four transforms are working and verified.

## 1. Patch Occlusion
* **Mechanism:** Zero out a random rectangular patch per `(sample, timestep, view)`.
* **Expected:** Mean pixel value drops proportionally to patch area fraction (5–50% of H×W).
* **Result:** Means drop 25–30% on bright views (low, side). *(Note: High baseline %zero on over/wrist reflects dark frames, not the op).*
```text
[patch] low:   mean 99.389 → 71.796   %zero=29.86%
[patch] over:  mean 28.154 → 19.002   %zero=71.04%
[patch] side:  mean 97.611 → 70.654   %zero=30.19%
[patch] wrist: mean 62.114 → 44.885   %zero=60.22%
```
## 2. View Drop
- **Mechanism**: Zero an entire view for some samples, flip its `pad_mask_dict` entry to `False`.
- **Expected**: Selected views are all-zero, corresponding pmd is all-False.
- **Result**:
```text
[view drop] low:   mean 99.389 → 0.000   %zero=100.00%
[view drop] over:  mean 28.154 → 0.000   %zero=100.00%
[view drop] side:  mean 97.611 → 0.000   %zero=100.00%
[view drop] wrist: mean 62.114 → 0.000   %zero=100.00%
[view_drop_pmd] {'low': False, 'over': False, 'side': False, 'wrist': False}
```
## 3. Image Key Shuffle
* **Mechanism:** Per-sample random permutation of which camera occupies which image-key slot.
* **Expected:** Each slot's mean converges to the batch-wide across-view average `(99.4 + 28.2 + 97.6 + 62.1) / 4 ~ 71.8`. This occurs because each slot now holds a uniform mix of all 4 cameras across the batch.
* **Result:** All four slots land within ~6 units of the 71.8 expected average. Also verified per-sample multiset preservation (no view duplication or loss).
```text
[shuffle] low:   mean 99.389 → 71.104   %zero=25.79%
[shuffle] over:  mean 28.154 → 77.461   %zero=23.97%
[shuffle] side:  mean 97.611 → 70.892   %zero=27.55%
[shuffle] wrist: mean 62.114 → 67.811   %zero=33.40%
```
## 4. Proprio Sample Drop
- **Mechanism**: Per sample, zero all proprio subkeys and flip all proprio pmd entries. This models total proprio sensor loss (per-key independence handled by token drop)
- **Expected**: Every proprio subkey's pmd fraction equals `1 - prob` (0 at `prob = 1.0`)
- **Result**:
```
[prop_drop_pmd] {
    'gripper':     0.0,
    'joints':      0.0,
    'orientation': 0.0,
    'position':    0.0,
}
```
## Implementation Note
All four methods reside in the Grain pipeline (pre-np2jax, train-only) and are gated by independent `*_prob` config flags that default to `0.0`.

# Model Transform
## `proprio_token_drop` (in model)
Wired through `LowdimObsTokenizer` via `p_token_drop`, surfaced on `ModelFactory.proprio_token_drop_prob` -> CLI as `--model.proprio-token-drop-prob`.

## Issue
**Had to re-enable proprio tokenizers to test**:
`scripts/train/xflow.py:257` hardcodes `cfg.model.proprio_keys = ()` which bypasses the `LowdimObsTokenizer` instantiation entirely, so `p_token_drop` silently does nothing in the default config. Verified via grep that no other model path currently consumes proprio (`DofTokenizer` is defined but never imported; `act_vae.py` is only referenced from `tests/broken`.)
To verify the token drop fired, swapped line 257 to:
```
cfg.model.proprio_keys = inferred_proprio_keys
```
This hits the per-subkey `LowdimObsTokenizer` instances built by `make_obs_proprio`, which is where `p_token_drop` runs. Need to discuss as leaving proprio out of pipeline skips over `proprio_sample_drop` and this token drop.

## Result with `--model.proprio-token-drop-prob 0.5`
Ran at prob=0.5 for ~40 steps, inspected `TokenGroup.mask.mean()` per subkey:
```
[observation_tokenizers_gripper]     mask.mean ≈ 0.51   shape=(256, 1, 1)
[observation_tokenizers_joints]      mask.mean ≈ 0.50   shape=(256, 1, 7)
[observation_tokenizers_orientation] mask.mean ≈ 0.50   shape=(256, 1, 3)
[observation_tokenizers_position]    mask.mean ≈ 0.49   shape=(256, 1, 3)
```
All four checks pass:
1. **Prob matches**: every subkey's `mask.mean` sits at ~0.5, confirming that the Bernoulli gate fires
2. **Shape is `(B, T, D)` per subkey**: one independent coin per (sample, timestep, DOF-slot). This is true per-token granularity, not per-subkey. A single sample can have 4 of 7 joints dropped while keeping all 3 position DOFs, exactly matching the partial sensor failure semantics as discussed.
3. **Subkey independence**: means drift independently step-to-step (e.g. position=0.47 while gripper-0.51 at step 35), confirming each tokenizer's `make_rng("dropout")` stream is uncorrelated. There is no rng collision.
4. **Mask, not value-zero**: `p_token_drop` flips the attention mask via `mask & keep`; the transformer skips dropped tokens rather than attending to zeroed values. Leaves the existing `dropout_rate=0.2` scalar drop in place.
